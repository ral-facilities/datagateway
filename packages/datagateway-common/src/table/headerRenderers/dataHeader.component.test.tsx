import React from 'react';
import { mount, shallow } from 'enzyme';
import DataHeader from './dataHeader.component';
import TextColumnFilter from '../columnFilters/textColumnFilter.component';
import { TableSortLabel } from '@mui/material';

describe('Data column header component', () => {
  const onSort = jest.fn();
  const resizeColumn = jest.fn();
  const dataHeaderProps = {
    label: 'Test',
    labelString: 'Test',
    dataKey: 'test',
    sort: {},
    onSort,
    resizeColumn,
    icon: function Icon() {
      return <div>Test</div>;
    },
  };

  const filterComponent = (
    label: string,
    dataKey: string
  ): React.ReactElement => (
    <TextColumnFilter label={label} onChange={jest.fn()} value={undefined} />
  );

  afterEach(() => {
    onSort.mockClear();
    resizeColumn.mockClear();
  });

  it('renders correctly without sort or filter', () => {
    const wrapper = shallow(
      <DataHeader {...dataHeaderProps} disableSort={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with sort but no filter', () => {
    const wrapper = shallow(<DataHeader {...dataHeaderProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with filter but no sort', () => {
    const wrapper = shallow(
      <DataHeader
        {...dataHeaderProps}
        disableSort={true}
        filterComponent={filterComponent}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with sort and filter', () => {
    const wrapper = shallow(
      <DataHeader {...dataHeaderProps} filterComponent={filterComponent} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('calls the onSort method when label is clicked', () => {
    it('sets asc order', () => {
      const wrapper = shallow(<DataHeader {...dataHeaderProps} />);

      const label = wrapper.find(TableSortLabel);

      label.simulate('click');
      expect(onSort).toHaveBeenCalledWith('test', 'asc', 'push');
    });

    it('sets desc order', () => {
      const wrapper = shallow(
        <DataHeader {...dataHeaderProps} sort={{ test: 'asc' }} />
      );

      const label = wrapper.find(TableSortLabel);

      label.simulate('click');
      expect(onSort).toHaveBeenCalledWith('test', 'desc', 'push');
    });

    it('sets null order', () => {
      const wrapper = shallow(
        <DataHeader {...dataHeaderProps} sort={{ test: 'desc' }} />
      );

      const label = wrapper.find(TableSortLabel);

      label.simulate('click');
      expect(onSort).toHaveBeenCalledWith('test', null, 'push');
    });
  });

  describe('calls the onSort method when default sort is specified', () => {
    it('sets asc order', () => {
      const wrapper = mount(
        <DataHeader {...dataHeaderProps} defaultSort="asc" />
      );
      wrapper.update();

      expect(onSort).toHaveBeenCalledWith('test', 'asc', 'replace');
    });

    it('sets desc order', () => {
      const wrapper = mount(
        <DataHeader {...dataHeaderProps} defaultSort="desc" />
      );
      wrapper.update();

      expect(onSort).toHaveBeenCalledWith('test', 'desc', 'replace');
    });
  });

  it('calls the resizeColumn method when column resizer is dragged', () => {
    const wrapper = shallow(<DataHeader {...dataHeaderProps} />);

    wrapper.find('Draggable').prop('onDrag')(null, { deltaX: 50 });

    expect(resizeColumn).toHaveBeenCalledWith('test', 50);
  });
});
