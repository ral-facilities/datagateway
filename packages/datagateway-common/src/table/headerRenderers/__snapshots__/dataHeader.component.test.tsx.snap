// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data column header component renders correctly with filter but no sort 1`] = `
<ForwardRef(TableCell)
  component="div"
  size="small"
  variant="head"
>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <ForwardRef(Box)
      display="flex"
    >
      <ForwardRef(Box)
        marginRight={1}
      >
        <Icon />
      </ForwardRef(Box)>
      <ForwardRef(Box)>
        <ForwardRef(Typography)
          noWrap={true}
          sx={
            Object {
              "fontSize": "inherit",
              "lineHeight": "inherit",
            }
          }
        >
          Test
        </ForwardRef(Typography)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
    <TextColumnFilter
      label="Test"
      onChange={[MockFunction]}
    />
  </div>
  <Draggable
    allowAnyClick={false}
    axis="none"
    bounds={false}
    cancel={null}
    defaultClassName="react-draggable"
    defaultClassNameDragged="react-draggable-dragged"
    defaultClassNameDragging="react-draggable-dragging"
    defaultPosition={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    position={null}
    scale={1}
    transform={null}
  >
    <div
      style={
        Object {
          "cursor": "col-resize",
          "marginLeft": 18,
          "paddingLeft": "4px",
          "paddingRight": "4px",
        }
      }
    >
      <ForwardRef(Divider)
        flexItem={true}
        orientation="vertical"
        sx={
          Object {
            "height": "100%",
          }
        }
      />
    </div>
  </Draggable>
</ForwardRef(TableCell)>
`;

exports[`Data column header component renders correctly with sort and filter 1`] = `
<ForwardRef(TableCell)
  component="div"
  size="small"
  variant="head"
>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <ForwardRef(Box)
      display="flex"
    >
      <ForwardRef(Box)
        marginRight={1}
      >
        <Icon />
      </ForwardRef(Box)>
      <ForwardRef(Box)>
        <ForwardRef(TableSortLabel)
          active={false}
          className="tour-dataview-sort"
          onClick={[Function]}
        >
          <ForwardRef(Typography)
            noWrap={true}
            sx={
              Object {
                "fontSize": "inherit",
                "lineHeight": "inherit",
              }
            }
          >
            Test
          </ForwardRef(Typography)>
        </ForwardRef(TableSortLabel)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
    <TextColumnFilter
      label="Test"
      onChange={[MockFunction]}
    />
  </div>
  <Draggable
    allowAnyClick={false}
    axis="none"
    bounds={false}
    cancel={null}
    defaultClassName="react-draggable"
    defaultClassNameDragged="react-draggable-dragged"
    defaultClassNameDragging="react-draggable-dragging"
    defaultPosition={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    position={null}
    scale={1}
    transform={null}
  >
    <div
      style={
        Object {
          "cursor": "col-resize",
          "marginLeft": 18,
          "paddingLeft": "4px",
          "paddingRight": "4px",
        }
      }
    >
      <ForwardRef(Divider)
        flexItem={true}
        orientation="vertical"
        sx={
          Object {
            "height": "100%",
          }
        }
      />
    </div>
  </Draggable>
</ForwardRef(TableCell)>
`;

exports[`Data column header component renders correctly with sort but no filter 1`] = `
<ForwardRef(TableCell)
  component="div"
  size="small"
  variant="head"
>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <ForwardRef(Box)
      display="flex"
    >
      <ForwardRef(Box)
        marginRight={1}
      >
        <Icon />
      </ForwardRef(Box)>
      <ForwardRef(Box)>
        <ForwardRef(TableSortLabel)
          active={false}
          className="tour-dataview-sort"
          onClick={[Function]}
        >
          <ForwardRef(Typography)
            noWrap={true}
            sx={
              Object {
                "fontSize": "inherit",
                "lineHeight": "inherit",
              }
            }
          >
            Test
          </ForwardRef(Typography)>
        </ForwardRef(TableSortLabel)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </div>
  <Draggable
    allowAnyClick={false}
    axis="none"
    bounds={false}
    cancel={null}
    defaultClassName="react-draggable"
    defaultClassNameDragged="react-draggable-dragged"
    defaultClassNameDragging="react-draggable-dragging"
    defaultPosition={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    position={null}
    scale={1}
    transform={null}
  >
    <div
      style={
        Object {
          "cursor": "col-resize",
          "marginLeft": 18,
          "paddingLeft": "4px",
          "paddingRight": "4px",
        }
      }
    >
      <ForwardRef(Divider)
        flexItem={true}
        orientation="vertical"
        sx={
          Object {
            "height": "100%",
          }
        }
      />
    </div>
  </Draggable>
</ForwardRef(TableCell)>
`;

exports[`Data column header component renders correctly without sort or filter 1`] = `
<ForwardRef(TableCell)
  component="div"
  size="small"
  variant="head"
>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <ForwardRef(Box)
      display="flex"
    >
      <ForwardRef(Box)
        marginRight={1}
      >
        <Icon />
      </ForwardRef(Box)>
      <ForwardRef(Box)>
        <ForwardRef(Typography)
          noWrap={true}
          sx={
            Object {
              "fontSize": "inherit",
              "lineHeight": "inherit",
            }
          }
        >
          Test
        </ForwardRef(Typography)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </div>
  <Draggable
    allowAnyClick={false}
    axis="none"
    bounds={false}
    cancel={null}
    defaultClassName="react-draggable"
    defaultClassNameDragged="react-draggable-dragged"
    defaultClassNameDragging="react-draggable-dragging"
    defaultPosition={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    position={null}
    scale={1}
    transform={null}
  >
    <div
      style={
        Object {
          "cursor": "col-resize",
          "marginLeft": 18,
          "paddingLeft": "4px",
          "paddingRight": "4px",
        }
      }
    >
      <ForwardRef(Divider)
        flexItem={true}
        orientation="vertical"
        sx={
          Object {
            "height": "100%",
          }
        }
      />
    </div>
  </Draggable>
</ForwardRef(TableCell)>
`;
